src/test/scala/org/apache/spark/sql/delta/MergeIntoSuiteBase.scala

```diff
@@ -1878,4 +1880,79 @@ abstract class MergeIntoSuiteBase
       (1, 10) // update
     )
   )
+
+  protected def withJobCountListener(jobCount: AtomicInteger)(f: => Unit): Unit = {
+    val listener = new SparkListener {
+      override def onJobStart(jobStart: SparkListenerJobStart): Unit = {
+        // Spark will always log a job start/end event even when the job does not launch any task.
+        if (jobStart.stageInfos.exists(_.numTasks > 0)) {
+          jobCount.incrementAndGet()
+        }
+      }
+    }
+    sparkContext.addSparkListener(listener)
+    sparkContext.listenerBus.waitUntilEmpty(15000)
+    jobCount.set(0)
+
+    Utils.tryWithSafeFinally(f) {
+      sparkContext.listenerBus.waitUntilEmpty(15000)
+      sparkContext.removeSparkListener(listener)
+    }
+  }
+
+  test("findTouchedFiles - no files match target predicate") {
+
+    withTable("source") {
+      append(Seq((1, 10), (2, 20)).toDF("key1", "value1"), "value1")  // target
+      val source = Seq((3, 30)).toDF("key2", "value2")  // source
+
+      val jobCount = new AtomicInteger(0)
+
+      withJobCountListener(jobCount) {
+        io.delta.tables.DeltaTable.forPath(spark, tempPath)
+          .merge(source, "key1 = key2 and value1 in (30)")
+          .whenMatched().updateExpr(Map("key1" -> "key2", "value1" -> "value2"))
+          .whenNotMatched().insertExpr(Map("key1" -> "key2", "value1" -> "value2"))
+          .execute()
+      }
+
+      assert(jobCount.get() == 3, "There should be 3 Spark jobs")
+
+      checkAnswer(
+        readDeltaTable(tempPath),
+        Row(1, 10) ::    // No change
+          Row(2, 20) ::     // No change
+          Row(3, 30) ::     // Insert
+          Nil)
+
+    }
+  }
+
+  test("findTouchedFiles - files match target predicate") {
+
+    withTable("source") {
```

Comment:
These two tests together essentially test just ONE thing - when there is no match the number of spark jobs is less. So I suggest merging these two tests together like this. 
```
test(...) {
   // append
   // val jobCountWithMatch = merge with source1
   // val jobCountWithoutMatch = merge with source2
   // assert(jobCountWithoutMatch < jobCountWithMatch)
}
```
This will make the test more robust and future-proof. Right now it is brittle with the assumption that there will be exactly 3 or 5 jobs. Rather if you ignore the abosolute counts, and just compare them relatively, then its much better.

Reply:
@tdas Makes a lot of sense. I asked @brkyvz about the non-deterministic nature of the job count, and I don't think I understood his response correctly.